File: api/src/dressme/wearable_classification.py
import io

from google import genai
from google.genai import types
from PIL import Image
from pydantic import BaseModel

from . import schemas


class ClassificationResult(BaseModel):
    category: schemas.WearableCategory


class WearableClassifier:
    def __init__(self, api_key: str):
        self._client = genai.Client(api_key=api_key)

    async def classify(self, image_data: bytes) -> schemas.WearableCategory:
        """Classify a wearable image into one of the known categories.

        Approximate cost: $0.0003 per invocation
        """
        image = Image.open(io.BytesIO(image_data))
        image.thumbnail((512, 512))

        response = await self._client.aio.models.generate_content(
            model="gemini-3.1-flash-lite-preview",
            contents=[image, "classify this wearable"],
            config=types.GenerateContentConfig(
                response_mime_type="application/json",
                response_schema=ClassificationResult.model_json_schema(),
            ),
        )

        if response.text is None:
            raise ValueError("Response text is None")

        result = ClassificationResult.model_validate_json(response.text)
        return result.category

File: src/server/services/wearable-classification.ts
import { GoogleGenAI, ApiError } from "@google/genai";
import pRetry from "p-retry";
import { getSettings } from "../settings";
import { WEARABLE_CATEGORIES, type WearableCategory } from "@/shared/wearable-categories";
import { classifyResponseSchema } from "@/shared/schemas";

async function getSharp() {
  return (await import("sharp")).default;
}

/**
 * Classify a wearable image into one of the known categories.
 * Approximate cost: $0.0003 per invocation.
 */
export async function classifyWearableImage(imageData: Buffer): Promise<WearableCategory | null> {
  const settings = getSettings();
  const ai = new GoogleGenAI({ apiKey: settings.GEMINI_API_KEY });

  const sharp = await getSharp();
  const downscaled = await sharp(imageData)
    .resize(512, 512, { fit: "inside", withoutEnlargement: true })
    .jpeg()
    .toBuffer();

  const response = await pRetry(
    () =>
      ai.models.generateContent({
        model: "gemini-3.1-flash-lite",
        contents: [
          { inlineData: { mimeType: "image/jpeg", data: downscaled.toString("base64") } },
          "classify this wearable",
        ],
        config: {
          responseMimeType: "application/json",
          responseSchema: {
            type: "OBJECT",
            properties: {
              category: {
                type: "STRING",
                nullable: true,
                description: `The category of the wearable, one of: ${WEARABLE_CATEGORIES.join(", ")}`,
              },
            },
          },
        },
      }),
    {
      retries: 3,
      shouldRetry: ({ error }) =>
        error instanceof TypeError ||
        (error instanceof ApiError && [408, 429, 500, 502, 503, 504].includes(error.status)),
      onFailedAttempt: ({ attemptNumber, retriesLeft, error }) => {
        console.info(
          `classifyWearableImage attempt ${attemptNumber} failed (${retriesLeft} retries left): ${error.message}`,
        );
      },
    },
  );

  const text = response.text;
  if (!text) throw new Error("Response text is null");

  const parsed = classifyResponseSchema.parse(JSON.parse(text));
  return parsed.category;
}
